nest-forms-backend/src/utils/subservices/sharepoint.subservice.ts

```diff
@@ -0,0 +1,357 @@
+    columns: Array<string>,
+    accessToken: string,
+    dtbName: string,
+  ): Promise<Record<string, string>> {
+    const result: Record<string, string> = {}
+    const { SHAREPOINT_URL } = process.env
+    const url = `${escape(SHAREPOINT_URL)}/lists/getbytitle('${dtbName}')/fields`
+
+    const fields = await axios
+      .get(url, {
+        headers: {
+          Accept: 'application/json;odata=verbose',
+          'Content-Type': 'application/json;odata=verbose',
+          Authorization: `Bearer ${accessToken}`,
+        },
+      })
+      .then(
+        (
+          response: AxiosResponse<
+            { d: { results: List<Record<string, string>> } },
+            object
+          >,
+        ) => response.data.d.results,
+      )
+
+    columns.forEach((col) => {
+      const filtered = filter<Record<string, string>>(
+        fields,
+        (field) => field.Title === col,
+      )
+      if (filtered.length === 0) {
+        throw this.throwerErrorGuard.BadRequestException(
+          SharepointErrorsEnum.UNKNOWN_COLUMN,
+          `${SharepointErrorsResponseEnum.UNKNOWN_COLUMN} Column: ${col}, dtb name: ${dtbName}.`,
+        )
+      }
+      result[col] = filtered[0].StaticName
+    })
+
+    return result
+  }
+
+  private async postDataToSharepoint(
+    dtbName: string,
+    accessToken: string,
+    fieldValues: Record<string, string>,
+  ): Promise<number> {
+    const { SHAREPOINT_URL } = process.env
+    const url = `${escape(SHAREPOINT_URL)}/lists/getbytitle('${dtbName}')/items`
+    const postData = {
+      ...fieldValues,
+    }
+
+    const result = await axios
+      .post(url, postData, {
+        headers: {
+          Accept: 'application/json;odata=verbose',
+          Authorization: `Bearer ${accessToken}`,
+        },
+      })
+      .catch((error) => {
+        throw this.throwerErrorGuard.BadRequestException(
+          SharepointErrorsEnum.POST_DATA_TO_SHAREPOINT_ERROR,
+          `${
+            SharepointErrorsResponseEnum.POST_DATA_TO_SHAREPOINT_ERROR
+          } Error: ${<string>error} when sending to database: ${dtbName}, posted data: ${JSON.stringify(postData)} .`,
+        )
+      })
+      .then(
+        (res: AxiosResponse<{ d: { ID: number } }, object>) => res.data.d.ID,
+      )
+
+    return result
+  }
+
+  private async getAccessToken(): Promise<string> {
+    const {
+      SHAREPOINT_TENANT_ID,
+      SHAREPOINT_CLIENT_ID,
+      SHAREPOINT_CLIENT_SECRET,
+      SHAREPOINT_DOMAIN,
+    } = process.env
+    const url = `https://accounts.accesscontrol.windows.net/${escape(
+      SHAREPOINT_TENANT_ID,
+    )}/tokens/OAuth/2`
+    const result = await axios
+      .post(
+        url,
+        {
+          grant_type: 'client_credentials',
+          client_id: `${escape(SHAREPOINT_CLIENT_ID)}@${escape(
+            SHAREPOINT_TENANT_ID,
+          )}`,
+          client_secret: SHAREPOINT_CLIENT_SECRET,
+          resource: `00000003-0000-0ff1-ce00-000000000000/${
```

Comment:
Shouldn't `00000003-0000-0ff1-ce00-000000000000` be in some env variable too?

Reply:
For some reason this is just some constant to configure sharepoint API :D https://learn.microsoft.com/en-us/sharepoint/hybrid/configure-server-to-server-authentication